devlib/instrument/schedstats.py

```diff
@@ -0,0 +1,166 @@
+    schedstat_path = '/proc/schedstat'
+
+    def setup(self):
+        try:
+            self.old_sysctl_value = self.target.read_int(self.sysctl_path)
+        except TargetError:
+            if not self.target.file_exists(self.sysctl_path):
+                raise TargetError('schedstats not supported by target. '
+                                  'Ensure CONFIG_SCHEDSTATS is enabled.')
+
+        self.target.write_value(self.sysctl_path, 1)
+
+        # Check version matches
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        match = re.search(r'version ([0-9]+)', lines[0])
+        if not match or match.group(1) != '15':
+            raise TargetError(
+                'Unsupported schedstat version string: "{}"'.format(lines[0]))
+
+        # Take a sample of the schedstat file to figure out which channels to
+        # create.
+        # We'll create a site for each CPU and a site for each sched_domain.
+        for site, measures in self._get_sample().iteritems():
+            if site.startswith('cpu'):
+                measurement_category = 'schedstat_cpu'
+            else:
+                measurement_category = 'schedstat_domain'
+
+            for measurement_name in measures.keys():
+                # TODO: constructing multiple MeasurementTypes with same
+                # params. Does it matter?
+                measurement_type = MeasurementType(
+                    measurement_name, '', measurement_category)
+                self.add_channel(site=site,
+                                 name='{}_{}'.format(site, measurement_name),
+                                 measure=measurement_type)
+
+    def teardown(self):
+        self.target.write_value(self.sysctl_path, self.old_sysctl_value)
+
+    def _get_sample(self):
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        ret = OrderedDict()
+
+        # Example /proc/schedstat contents:
+        #
+        # version 15
+        # timestamp <timestamp>
+        # cpu0 <cpu fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+        # cpu1 <cpu_fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+
+        curr_cpu = None
+        for line in lines[2:]:
+            tokens = line.split()
+            if tokens[0].startswith('cpu'):
+                curr_cpu = tokens[0]
+                site = curr_cpu
+                measures = CPU_MEASURES
+            elif tokens[0].startswith('domain'):
+                if not curr_cpu:
+                    raise TargetError(
+                        'Failed to parse schedstats, found domain before CPU')
+                # We'll name the site for the domain like "cpu0domain0"
+                site = curr_cpu + tokens[0]
+                measures = DOMAIN_MEASURES
+            else:
+                raise TargetError(
+                    'Unrecognised schedstats line: "{}"'.format(line))
```

Comment:
This will fails in all our Android systems, since we have additional stats reporting EAS codepaths.
In Android kernel we will have stats like:

```
cpu0 9703 0 227370 45501 209763 87404 75510237990 76927923766 168389
eas 50510 0 0 29148 0 21362 51469 11340 3022 43 14180 169 22712 40129 3 0 3023 28011 15055 11640
domain0 0f 5237 5089 86 95279 66 0 0 5089 2021 1945 62 74554 19 1 3 1942 43224 36754 5499 4393208 971 0 7128 29626 1 0 1 0 0 0 0 0 0 39102 1466 0
eas 0 0 0 9389 0 0 0 0 0 0 0 0 0 0 0 0 0 0 13480 0
domain1 ff 3981 3733 220 673383 33 2 5 1149 1076 1069 6 36930 2 0 3 176 42131 36952 4441 18425660 738 3 261 36691 1 0 1 0 0 0 0 0 0 83215 794 0
eas 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 15055 0
```

Where basically, after each `cpu` and `domain` we do have an `eas` entry with EAS specific counters.

I was wondering if we can end up with a more "generic" parser which allows to accept custom defined lines.

Maybe we can use a "map" to define what are the columns for each specific "channel", the map can be defined by default using the hardcoded values you have now... but we can allow the user to specify a custom map.
In that case, the custom map should allow to parse additional fields (e.g. eas stats) as well as open the door to parse stats versions not supported by default.

What do you think?